### PR TITLE
[feat] Optimize TRTLLM Sampler to use pure python type for decoder buffer logits.

### PR DESCRIPTION
**Pull request by dcampora**
# Optimize TRTLLM Sampler to use pure python type for decoder buffer logits.

This PR avoids some C++ types where possible.

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/make_decoding_batch_input_output.py [hunk 1]
```diff
@@ -83,27 +83,25 @@ def create_decoder_batch_inputs(
...
 
         Args:
             context_requests: List of context requests
             generation_requests: List of generation requests
-            decoder_buffers: Decoder buffers
+            decoder_buffer_logits: Decoder buffer logits
             decoder_input_buffers: Decoder input buffers
             decoder_state: Current decoder state
-            model_config: Model configuration
             max_num_sequences: Maximum number of sequences to process
```
Comment (stnie) [inline tensorrt_llm/_torch/pyexecutor/make_decoding_batch_input_output.py:104]: Add a line for cache_indirections to the Docstring

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 2]
```diff
@@ -538,14 +531,27 @@ def _initialize_store(self):
...
-            "sequence_lengths_host":
-            torch.empty((
-                self.executor_config.max_batch_size,
-                self.executor_config.max_beam_width,
-            ),
-                        dtype=torch.int)
+            "decoder_state":
+            DecoderState(dtype=self.logits_datatype,
+                         buffer_manager=buffer_manager),
+            "cache_indirections": [
```
Comment (stnie) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:537]: cache_indirections should be of shape [**max_batch_size**, max_beam_width, max_attention_window]

Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:537]: Done.

**Review by QiJune — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 3]
```diff
@@ -612,21 +608,21 @@ def sample_async(self, scheduled_requests: ScheduledRequests,
             num_context_logits[
                 batch_index] = request.context_chunk_size if request.py_return_context_logits else 1
 
-        logits_index = self.algs.handle_context_logits(
+        decoder_buffer_logits, logits_index = self.algs.handle_context_logits(
             scheduled_requests.context_requests, num_context_logits,
-            model_outputs["logits"], self.store["decoder_buffers"])
+            model_outputs["logits"], self.executor_config.max_batch_size)
```
[the remark this answers is not in the document]
Comment (dcampora) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:613]: Done.
